### PR TITLE
Batch computation for OpenCL

### PR DESCRIPTION
**Pull request by frpays**
Ported the work of Henrik Forstén from GCP Leela-zero to lc0.

Based on https://github.com/LeelaChessZero/lc0/issues/115.
Based on https://github.com/Ttl/leela-zero/tree/batching_test.

Resolving https://github.com/LeelaChessZero/lc0/issues/115.

GTX Titan benchmarks:

Threads | master | batch_size=1 | batch_size=2 | batch_size=4 | batch_size=8
-- | -- | -- | -- | -- | --
1 | 798 | 797 | 1017 | 1158 | 1258
2 | 1228 | 1175 | 1350 | 1457 | 1511
4 | 1459 | 1463 | 1624 | 1598 | 1569
6 | 1411 | 1379 | 1665 | 1588 | 1571
8 | 1329 | 1316 | 1649 | 1587 | 1580

### CODE REVIEW

**Comment by Ttl** [comment]
How does the performance change if you tune the SGEMM for larger batch size? This can be done easily by multiplying `WINOGRAD_P` by the batch size in tuner call on `OpenCL.cc:766` if I'm reading the code correctly. It should give some performance gain especially on high end GPUs where the default tile size in SGEMM is too small to properly fully utilize it.

**Comment by frpays** [comment]
@Ttl Thank you very much for this remark. I will definitely investigate your suggestion.
First thing in order, I forgot to include the batching to the innerproduct/xgemv kernel.

**Comment by frpays** [comment]
I made the OpenCL batch_size a runtime parameter. The default is 2, but I think I will have to change it back to 1, as maybe no all devices may even support 2.

**Comment by frpays** [comment]
@Ttl It's about 5% to 10% faster with a sgemm batch_size of 4x16=64 on my GTX Titan.

File: src/neural/network_opencl.cc [hunk 1]
```diff
@@ -59,72 +59,97 @@ class OpenCLComputation : public NetworkComputation {
...
-    for (const InputPlane& plane : sample) {
-      float value = plane.value;
-      const uint64_t one = 1;
-      for (int i = 0; i < 64; i++)
-        input_data_[index++] = (plane.mask & (one << i)) != 0 ? value : 0;
-    }
+    const auto num_output_policies = weights_.num_output_policies;
+    const auto num_value_channels = weights_.num_value_channels;
+
+    /* Typically
```
Comment (mooskagh) [inline src/neural/network_opencl.cc]: We use `//` style comments, not `/* */`.

File: src/neural/network_opencl.cc [hunk 2]
```diff
@@ -59,72 +59,97 @@ class OpenCLComputation : public NetworkComputation {
...
-    std::vector<float> policy_data(weights_.num_output_policies);
-    opencl_net_.forward(input_data_, policy_data, value_data_);
+      for (size_t j = 0; j < batch_size; j++) {
+        std::vector<float> policy(weights_.num_output_policies);
 
-    // Get the moves
-    FullyConnectedLayer::Softmax(weights_.num_output_policies,
-                                 policy_data.data(), policy_data.data());
-    policy_data_.emplace_back(move(policy_data));
+        // Get the moves
```
Comment (mooskagh) [inline src/neural/network_opencl.cc]: Dot at the end of comments.

File: src/neural/network_opencl.cc [hunk 3]
```diff
@@ -59,72 +59,97 @@ class OpenCLComputation : public NetworkComputation {
...
-    // Now get the score
-    double winrate = FullyConnectedLayer::Forward0D(weights_.num_value_channels,
-                                                    weights_.ip2_val_w.data(),
-                                                    value_data_.data()) +
-                     weights_.ip2_val_b[0];
-    q_value_.emplace_back(std::tanh(winrate));
+        policies_.emplace_back(std::move(policy));
+
+        // Now get the score
+        double winrate = FullyConnectedLayer::Forward0D(
```
Comment (mooskagh) [inline src/neural/network_opencl.cc]: float

File: src/neural/network_opencl.cc [hunk 4]
```diff
@@ -59,72 +59,97 @@ class OpenCLComputation : public NetworkComputation {
...
-  std::vector<float> q_value_;
+  std::vector<std::vector<float>> policies_;
+  std::vector<float> q_values_;
 };
 
+void OpenCLComputation::EncodePlanes(const InputPlanes& sample, float* buffer) {
+  for (const InputPlane& plane : sample) {
+    const float value = plane.value;
+    for (auto i = 0; i < kSquares; i++)
+      *(buffer++) = (plane.mask & (((uint64_t)1) << i)) != 0 ? value : 0;
```
Comment (mooskagh) [inline src/neural/network_opencl.cc]: Curly braces around the statement, as it's in a separate line.

File: src/neural/network_opencl.cc [hunk 5]
```diff
@@ -137,6 +162,16 @@ class OpenCLNetwork : public Network {
...
         options.GetOrDefault<bool>("tune_exhaustive", false);
+    params_.tune_batch_size =
+        options.GetOrDefault<int>("tune_batch_size", 2);
+
+    auto max_batch_size_ =
+        static_cast<size_t>(options.GetOrDefault<int>("batch_size", 1));
+    if (max_batch_size_ > kHardMaxBatchSize) {
+      max_batch_size_ = kHardMaxBatchSize;
+    }
+    fprintf(stderr, "OpenCL, maximum batch size set to %ld.\n",
```
Comment (mooskagh) [inline src/neural/network_opencl.cc]: std::cerr

File: src/neural/network_opencl.cc (hunk 1 above)
Comment (frpays) [reply src/neural/network_opencl.cc]: Maybe that should start with the license statement ?

**Comment by frpays** [comment]
All change requests addressed.

**Comment by gsobala** [comment]
Very impressive speedup on macos / AMD Vega 56: 5x ! Gets 1816 nps with batch_size=12/threads=2

**Pull request merged**